### PR TITLE
ADCP the analog input multiplexer

### PR DESCRIPTION
**Pull request by Ceiku**
## Description:
New feature that multiplex access to an analog pin, or equivalent by powering sensors from digital output pins. 
This is only a partial MVP that takes a single adc sensor, future commits will let you group to different adc inputs.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/adcp/adcp.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  void update() override;
+
+ protected:
+  ADCPComponent *parent_;
+  GPIOPin *power_pin_;
+  uint8_t delay_;
+};
+
+
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.h:42]:
```suggestion
```

File: esphome/components/adcp/adcp.cpp [hunk 2]
```diff
@@ -13,6 +13,7 @@ void ADCPSensor::update() {
   delay(this->delay_);
   float state = this->parent_->get_source()->sample();
   this->power_pin_->digital_write(false);
+  this->publish_state(state);
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.cpp:16]: Not sure how I missed publishing the state after setting the float..haha

File: esphome/components/adcp/adcp.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+void ADCPComponent::dump_config() { ESP_LOGCONFIG(TAG, "Setting up ADCP..."); }
+
+void ADCPSensor::update() {
+  this->power_pin_->digital_write(true);
+  delay(this->delay_);
+  float state = this->parent_->get_source()->sample();
+  this->power_pin_->digital_write(false);
+  this->publish_state(state);
+}
+
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.cpp:18]:
```suggestion
void ADCPSensor::update() {
  this->power_pin_->digital_write(true);
  this->read_timestamp_ = millis();
  this->read_ = true;
}

void ADCPSensor::loop() {
  if (this->read_ && millis() > this->read_timestamp_ + this->delay_) {
    float state = this->parent_->get_source()->sample();
    this->power_pin_->digital_write(false);
    this->publish_state(state);
    this->read_ = false;
  }
}
```

File: esphome/components/adcp/adcp.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+class ADCPSensor : public sensor::Sensor, public PollingComponent {
+ public:
+  ADCPSensor(ADCPComponent *parent) : parent_(parent) {}
+  void set_power_pin(GPIOPin *power_pin) { this->power_pin_ = power_pin; }
+  void set_delay(uint8_t delay) { this->delay_ = delay; }
+
+  uint8_t get_delay() const { return this->delay_; }
+
+  void update() override;
+
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.h:35]:
```suggestion
  void update() override;
  void loop() override;

```

File: esphome/components/adcp/adcp.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ protected:
+  ADCPComponent *parent_;
+  GPIOPin *power_pin_;
+  uint8_t delay_;
+};
+
+
+}  // namespace adcp
+}  // namespace esphome
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.h:44]:
```suggestion
  uint8_t delay_;
  uint32_t read_timestamp_{0};
  bool read_{false};
};

}  // namespace adcp
}  // namespace esphome
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Actually a better way to do this would be not to delay in `update()` because I think it holds up the whole ESPHome application.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# ARG BUILD_FROM=esphome/esphome-base-amd64:2.6.0
```
Comment (OttoWinter) [inline Dockerfile:1]: Please remove this change (and docker-compose.yml). Also, the `config` folder is excluded in `.gitignore`, so you can store your dev setup there.

File: esphome/components/adcp/adcp.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+namespace esphome {
+namespace adcp {
+
+static const char *TAG = "ADCP";
+
+void ADCPComponent::dump_config() { ESP_LOGCONFIG(TAG, "Setting up ADCP..."); }
+
+void ADCPSensor::update() {
+  this->power_pin_->digital_write(true);
+  delay(this->delay_);
```
Comment (OttoWinter) [inline esphome/components/adcp/adcp.cpp:13]: Hmm, this delay is pretty long (100ms by default) and will block the main esphome loop, please use `this->set_timeout` instead.

File: esphome/components/adcp/adcp.cpp (hunk 2 above)
Comment (Ceiku) [reply esphome/components/adcp/adcp.cpp:16]: Happens to the best of us, hehe!

File: esphome/components/adcp/adcp.h (hunk 5 above)
Comment (Ceiku) [reply esphome/components/adcp/adcp.h]: For some reason this got marked as outdated, I'll hardcopy your code over.

File: esphome/components/adcp/adcp.cpp (hunk 2 above)
Comment (Ceiku) [reply esphome/components/adcp/adcp.cpp]: Following @OttoWinter with a scheduler, which is better than making our own fixed time loop.
```
void ADCPSensor::update() {
  this->power_pin_->digital_write(true);
  this->set_timeout(this->delay_, this->complete_update);
}

void ADCPSensor::complete_update() {
  float state = this->parent_->get_source()->sample();
  this->power_pin_->digital_write(false);
  this->publish_state(state);
}
```

File: esphome/components/adcp/adcp.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+namespace esphome {
+namespace adcp {
+
+static const char *TAG = "ADCP";
+
+void ADCPComponent::dump_config() { ESP_LOGCONFIG(TAG, "Setting up ADCP..."); }
+
+void ADCPSensor::update() {
+  this->power_pin_->digital_write(true);
+  this->set_timeout((uint32_t) this->delay_, this->complete_update());
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.cpp:13]:
```suggestion
  this->set_timeout((uint32_t) this->delay_, [this](){ this->complete_update_(); });
```

File: esphome/components/adcp/adcp.cpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+void ADCPSensor::complete_update() {
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.cpp:16]:
```suggestion
void ADCPSensor::complete_update_() {
```

File: esphome/components/adcp/adcp.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  void set_power_pin(GPIOPin *power_pin) { this->power_pin_ = power_pin; }
+  void set_delay(uint8_t delay) { this->delay_ = delay; }
+
+  uint8_t get_delay() const { return this->delay_; }
+
+  void update() override;
+  void complete_update();
+
+ protected:
+  ADCPComponent *parent_;
```
Comment (jesserockz) [inline esphome/components/adcp/adcp.h:38]:
```suggestion

 protected:
  void complete_update_();
  
  ADCPComponent *parent_;
```

**Pull request closed without merging**